Update Im2Col related to make preparation for group conv in NHWC order.

xiaomengy opened the pull request:
Summary: Update Im2Col related to make preparation for group conv in NHWC order.

Differential Revision: D9285344

BIT-silence commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

houseroad commented on caffe2/utils/math_cpu.cc:
```diff
@@ -2519,113 +2807,57 @@ void Im2Col<float, CPUContext, StorageOrder::NCHW>(
-      }
-    }
+    CPUContext* context) {
+  // Fast conv with no dilation.
```
actually is zero padding and no dilation

houseroad commented on caffe2/utils/math_utils.h:
```diff
@@ -40,6 +40,18 @@ MATH_UTILS_DECL T Cube(const T x) {
+// negative value of a parameter converts it to value higher than
+// 0x800...
+// The casting allows to use one condition instead of two.
+MATH_UTILS_DECL bool IsAGeZeroAndALtB(const int a, const int b) {
```
Can we have a better name for this method? A little bit hard to read.

houseroad reviewed (approved):
LGTM, no need to do the loop on the group, which is great.

jspark1105 commented:
@BIT-silence thanks for the PR. This should help what I'm working on a lot. One comment. Using batched gemm for group convolution in NHCW definitely increases parallelism a lot, but at the same time it increases working set size (we're generating G times bigger col_buffer) so I think we should check how it affects the performance.

jspark1105 commented on caffe2/utils/math.h:
```diff
@@ -610,19 +610,19 @@ void CopyMatrix(
+    const int M,
+    const int N,
+    const T* A,
+    const int A_outer_stride,
```
Add comments on A/B_outer/inner_stride ?

BIT-silence replied:
I think the actual impact of the memory part will be less than we estimated. Since currently in the inference part we are using shared col_buffer, so the peek usage of that part will depends on the largest layer. If the whole model has at least one layer with G = 1, the actually usage of memory will not increase a lot.

jspark1105 replied:
Sorry that I was not clear. I was mainly talking about cache locality rather than the maximum col_buffer size.

xiaomengy committed:
Update Im2Col related to make preparation for group conv in NHWC order. (#10439)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10439

Update Im2Col related to make preparation for group conv in NHWC order.

Reviewed By: houseroad

Differential Revision: D9285344

fbshipit-source-id: eb4c581daf3e95050de6d8b1bdaf567164e7e1cd
